### PR TITLE
Limit bundle names and release names to 53 characters

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
## Problem
Release names for bundles deployed by fleet must not exceed 53 characters, however the current `SafeConcatName` function used is only limiting these values to 63 characters or less. In scenarios where the release name exceeds 53 the bundle will fail to deploy. 

## Solution
Limit bundle names and release names to 53 characters at most, this conforms with the fleet naming requirements.
 
## Testing
1. Happy path testing
+ Create an RKE2/K3s cluster with a very long name, such as `haffel-test-very-long-name-for-edge-cases-and-stuff`
+ run `kubectl get bundles.fleet.cattle.io -A` and ensure that all bundles are less than or equal to 53

2. Rancher Upgrade Testing
+ Create an RKE2/K3s cluster on an older version of Rancher where this issue is not seen
+ Wait for the cluster to become available 
+ Watch the bundles using `watch -n 1 kubectl get bundles.fleet.cattle.io -A`
+ Upgrade Rancher
+ Ensure that the bundles names are shortened and no bundles are left behind / not-modified

**Note:** This same issue occurs with the fleet-agent, this problem is being addressed in rancher/fleet here: https://github.com/rancher/fleet/pull/1353

### Manual Testing
Ive done the above

### Automated Testing
n/a

## QA Testing Considerations
Keep in mind https://github.com/rancher/fleet/pull/1353 contributes to this issue and blocks the proper roll out of the other bundles created by Rancher, 
 
### Regressions Considerations
We should make sure that existing clusters which do not need bundle name truncation do not receive it, so as to prevent unneeded redeployment of the bundles

### CODE REVIEW

**Review by snasovich — CHANGES_REQUESTED** [review]
LGTM, but I think we want to introduce a const (probably in already existing `pkg/fleet/const.go`) and reference it in all places. 53 is quite a magic number.

File: pkg/controllers/provisioningv2/managedchart/managedchart.go [hunk 1]
```diff
@@ -92,15 +92,15 @@ func (h *handler) OnChange(mcc *v3.ManagedChart, status v3.ManagedChartStatus) (
 
 	bundle := &v1alpha1.Bundle{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      name.SafeConcatName("mcc", mcc.Name),
+			Name:      name.Limit(name.SafeConcatName("mcc", mcc.Name), 53),
```
Comment (snasovich) [inline pkg/controllers/provisioningv2/managedchart/managedchart.go:95]: Also, do we have to shorten the Name itself (vs. ReleaseName). The change that introduced the limit doesn't seem to apply to Name (https://github.com/rancher/fleet/pull/964/files#diff-fe6024d9cdd994ea2d530051f203572ee902f5af2936f095b2bfd4f81f4aca26) but I may be missing something.

Comment (HarrisonWAffel) [reply pkg/controllers/provisioningv2/managedchart/managedchart.go]: I believe the release name is derived from the name in cases where one is not explicitly provided, which is why I shorted the name too. This was the same change made for the SUC bundle and wanted to cover both cases

**Comment by snasovich** [comment]
> LGTM, but I think we want to introduce a const (probably in already existing `pkg/fleet/const.go`) and reference it in all places. 53 is quite a magic number.

Just to close the loop - the reason it's not in `pkg/fleet/const.go` as the limit is not Fleet-enforced but rather Helm. https://github.com/helm/helm/issues/6006

**Review by snasovich — APPROVED** [review]

File: pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go [hunk 2]
```diff
@@ -6,6 +6,7 @@ import (
 	"encoding/hex"
 	"encoding/json"
 	"fmt"
+	"github.com/rancher/rancher/pkg/controllers/provisioningv2/rke2"
```
Comment (Oats87) [inline pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go:9]: I don't think this will pass the go imports validation

**Review by Oats87 — CHANGES_REQUESTED** [review]

File: pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go (hunk 2 above)
Comment (HarrisonWAffel) [reply pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go:9]: Okay, fixed in both this PR and the backport

**Review by snasovich — DISMISSED** [review]

File: pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go [hunk 3]
```diff
@@ -124,7 +125,7 @@ func (h *handler) OnChange(cluster *rancherv1.Cluster, status rancherv1.ClusterS
 	result = append(result, &v1alpha1.Bundle{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: cluster.Namespace,
-			Name:      name.SafeConcatName(cluster.Name, "managed", "system", "agent"),
+			Name:      name.Limit(name.SafeConcatName(cluster.Name, "managed", "system", "agent"), rke2.MaxHelmReleaseNameLength),
```
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go:128]: QQ, if we have a cluster name that is >= 53 characters, won't we end up creating identical bundles?

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go (hunk 3 above)
Comment (HarrisonWAffel) [reply pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go:128]: @jakefhyde  As in duplicate names for both the SUC and fleet-agent bundle? The additional strings we pass to `SafeConcatName` would ensure those two bundles don't share a name, though I might be misunderstanding your question

Comment (jakefhyde) [reply pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go:128]: If it gets truncated to 53 characters, and you can have a 64 character cluster name, then any cluster with the first 53 matching characters of another cluster will fail.

Comment (jakefhyde) [reply pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go:128]: ACtually should be fine, looks like `SafeConcatName` and `name.Limit` append the digest of the whole string, not the truncated version.

**Review by jakefhyde — DISMISSED** [review]

**Review by Oats87 — APPROVED** [review]

**Comment by HarrisonWAffel** [comment]
I've tested the upgrade scenario and confirmed that no bundles are orphaned

File: pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go (hunk 3 above)
Comment (HarrisonWAffel) [reply pkg/controllers/provisioningv2/rke2/managesystemagent/managesystemagent.go]: Just to follow up, this PR has been changed to no longer use a combination of wranglers `SafeConcatName` and `name.Limit` as it resulted in a double hashing of the string. A few function was added in Rancher and should be ported into Wrangler when we get the opportunity

**Review by snasovich — APPROVED** [review]

**Comment by snasovich** [comment]
Merging while @HarrisonWAffel as it's largely the same as [v2.6 PR](https://github.com/rancher/rancher/pull/40729) (for some reason there are minor differences in comments related to `SafeConcatName` func 🤷 ).
